### PR TITLE
Add design of switching kernel

### LINKED ISSUE #6549: auto data sync between different devices
project: https://github.com/PaddlePaddle/Paddle/issues/6403
In a sequence of operators, when some of them run on GPU, some can only run on CPU,  we need to auto add some operators to copy host memory and device memory, so the whole graph can run in a multi devices environment.

## some problem 
1.  now we use name to get var in scope, and data on different devices should have different names, if we add some operator to sync data between different devices, we should change the input name of corresponding operators.

### PR DESCRIPTION
**Pull request by reyoung**
Fix #6549

Also fix the model parallel and the MKLDNN support.

### CODE REVIEW

File: doc/design/switch_kernel.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  // For the most operators, the expected kernel type is as same as
+  // actual kernel type.
+  //
+  // So we pass `actual_kernel_type` as a parameter of 
+  // GetExpectedKernelType
+  auto expect_kernel_type = GetExpectedKernelType(actual_kernel_type);
+  
+  auto trans = g_data_transformation_[{actual_kernel_type, expect_kernel_type}];
+  
+  kenerl.run(trans(inputs));
```
Comment (typhoonzero) [inline doc/design/switch_kernel.md]: Typo "kenerl"

File: doc/design/switch_kernel.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
+
+We can infer a kernel type from the inputs of an operators. We let this kernel type as `actual kernel type`, which means this kernel type is the actually kernel type that operator should be performed.
+
+We can get a kernel type by 1) The configuration of operator description. (Users may want to force use `MKL` for `conv` operator). 2) The place of the current executor. (Executor is running on GPU). This kernel type is what we expect the operator will be performed on. We let this kernel type as `expect kernel type`.
```
Comment (typhoonzero) [inline doc/design/switch_kernel.md]: Question: Thought data layout is determined by the actual data feed to the operator, which is the user should care. If `MKL` use different ones, let the kernel transform it just in the `MKL` kernel, will this work?

File: doc/design/switch_kernel.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
+## Background
+Every operator has many kernels. It because there are multiple data types, places, data layout that Fluid supports. We use the `KernelType` to describe kernel types that operators can hold. 
+
+The `KernelType` is as follows.
+
+```
+struct KernelType {
+  Place place_;
+  DataType data_type_;
```
Comment (typhoonzero) [inline doc/design/switch_kernel.md]: Some operators have 2 types, like cost and evaluators, they must have label input which is sometimes int64.

File: doc/design/switch_kernel.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  LayoutType layout_;
+};
+```
+
+The `place_` is a descriptor of the device and the computational library, e.g., `MKLPlace`, `CUDAPlace`.
```
Comment (tensor-tang) [inline doc/design/switch_kernel.md]: Actually it's `MKLDNNPlace` in current code.

File: doc/design/switch_kernel.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+The `data_type_` is the data type that this kernel performs on, e.g., `FP32`, `INT64`. Note that one kernel may have inputs with different data types. However, it will be a major `data_type`. For example, the `cross_entropy` takes `int64` as it label, and `double`/`float` as its input logit and output cost. The major `data_type` of `cross_entropy` is `float`/`double`.
+
+The `layout` is useful for some computational library. One example is that MKLDNN uses many kinds of layout, such as `nChw8`. Each kind of layout will invoke the different kernel.
```
Comment (tensor-tang) [inline doc/design/switch_kernel.md]: It's `nChw8c`, not `nChw8`.
The `8c` means every 8 channels.

File: doc/design/switch_kernel.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  vec<Tensor> inputs = ...
+  auto actual_kernel_type = GetActualKernelType(inputs);
+  
+  // The expected kernel type is related to actual kernel type.
+  // For the most operators, the expected kernel type is as same as
+  // actual kernel type.
+  //
+  // So we pass `actual_kernel_type` as a parameter of 
+  // GetExpectedKernelType
+  auto expect_kernel_type = GetExpectedKernelType(actual_kernel_type);
```
Comment (tensor-tang) [inline doc/design/switch_kernel.md]: Maybe `kernel.GetExpectedKernelType` ?

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (tensor-tang) [reply doc/design/switch_kernel.md]: How can we get the expected kernel type and what determines the expected type?

File: doc/design/switch_kernel.md (hunk 1 above)
Comment (tensor-tang) [inline doc/design/switch_kernel.md]: How about the `Tensor` after converted ?  Would it be deleted? Or should stored at somewhere in case of re-alloc?

File: doc/design/switch_kernel.md (hunk 3 above)
Comment (dzhwinter) [reply doc/design/switch_kernel.md]: The below `major data_type` has considered this situation.

File: doc/design/switch_kernel.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed in GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
```
Comment (QiJune) [inline doc/design/switch_kernel.md]: The inputs of an operator are variables(Tensors). Do you mean that both operator and variable(Tensor) will have kernel type?

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (QiJune) [reply doc/design/switch_kernel.md]: So, the configuration of operator description need to be exposed to users and be set in ProgramDesc?

Comment (reyoung) [reply doc/design/switch_kernel.md]: > If MKL uses different ones, let the kernel transform it just in the MKL kernel, will this work?
It is not efficient.
There are many operators will be implemented by MKL.  `OpA --> OpB --> OpC`. If `OpB` and `OpC` use the MKL optimized layout format, we just transform data in `OpB`.

Comment (reyoung) [reply doc/design/switch_kernel.md]: > How can we get the expected kernel type and what determines the expected type?
Actually, I mean is this type really expected?

The `expected kernel type` can be determined by

1) Directly configured by users in the python program. It could be attributes of operators. e.g., There could be `force_cpu` in some operators and `use_mkldnn` in some operators. 

   We can give a unique way to configure kernel type for every kernel. It could be decided later since it does not block the work in this PR. Either an attribute named `use_mkldnn` or an unify attribute named `kernel type` is both okay.

2)

Comment (reyoung) [reply doc/design/switch_kernel.md]: > So, the configuration of operator description need to be exposed to users and be set in ProgramDesc?

We can give a unique way to configure kernel type for every kernel. It could be decided later since it does not block the work in this PR. Either an attribute named use_mkldnn or a unify attribute named kernel type are both okay.

File: doc/design/switch_kernel.md (hunk 6 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: Every operator instance has an expected kernel type. It is not an attribute in the kernel instance.

File: doc/design/switch_kernel.md (hunk 1 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: Well. It depends.
We only transform input data. We can re-alloc the memory and do not worry about the synchronize between the original tensor and the re-alloced tensor since it is the input of the operator and it is read-only in the operator.

We can delete the original data if the data is not used when this operator is complete.  However, I think reusing memory is another topic. We can archive that by tracing every variable's usage.

File: doc/design/switch_kernel.md (hunk 7 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: No. Only the operator has kernel type.

However, I think the operator has two kinds of kernel type. One is decided by user inputs. The other is decided by user configurations.

File: doc/design/switch_kernel.md (hunk 3 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: Yes. We have some operators which take two or more than two kinds of data types as their inputs.

However, I think it is easy to use a single data type for each operator. Even for the `cost` and `evaluators`, each of them has a single data type for output. And the output data type of them is decided by one input.

For example, `cross_entropy` has `logit` and `label` as its inputs, and produces `loss` as its output. The data type of `loss` is same as `logit`. So we think the data type of cross_entroy is `float` when `logit` is `float`.

The `label` can not only be `int64`, but also can be `int32`. We can add a manually switch in kernel object to achieve this.

File: doc/design/switch_kernel.md (hunk 1 above)
Comment (dzhwinter) [reply doc/design/switch_kernel.md]: I just want to confirm that I didn't misunderstand on this question.

 `Tensor` in different `data_transformation` function behave different.  There is 3 kind of situations.
For layout transform, `nChw8c ` => `NCHW`, we have to re-alloc a memory and assign the new Tensor with correct layout. 
For Datatype transform, `int`  => `float`, just reuse the Tensor.
For Place transform, if we support `Place Fallback`(I mean, will try to use the kernel by order <CUDNNconv, gemmGPU, gemmCPU >).  Do we allow GPUPlace fallback to CPUPlace? if so, we also need to re-alloc a CPU memory and make a copy.

Since the all the conversion will happen in every training instance, so the old-Tensor and new-Tensor must be re-used.

File: doc/design/switch_kernel.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+struct KernelType {
+  Place place_;
+  DataType data_type_;
+  LayoutType layout_;
+};
+```
+
+The `place_` is a descriptor of the device and the computational library, e.g., `MKLPlace`, `CUDAPlace`.
+
+The `data_type_` is the data type that this kernel performs on, e.g., `FP32`, `INT64`. Note that one kernel may have inputs with different data types. However, it will be a major `data_type`. For example, the `cross_entropy` takes `int64` as it label, and `double`/`float` as its input logit and output cost. The major `data_type` of `cross_entropy` is `float`/`double`.
```
Comment (jacquesqiao) [inline doc/design/switch_kernel.md]: do we need a uniform way to set which input is used to determine the major `data_type`, this can be done in `AddInput` of OpProtoAndCheckerMaker 
like `AsKernelTypeDecider`

File: doc/design/switch_kernel.md (hunk 1 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: > For Datatype transform, int => float, just reuse the Tensor.

**Reusing memory is another topic.**

We cannot reuse Tensor if the tensor is input data of other operators. Reusing memory should take care of the whole neural network. If the input is not used by other operators after the current operator, we can free the memory of this tensor.

We can give a quick implementation. The simplest way to implement the transformation of the tensor is always creating a new tensor.

There is no conflict between creating a new tensor while transforming and reusing memory. They are two independent topics. We can reuse memory even we create a new tensor every mini-batch.

File: doc/design/switch_kernel.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
+## Background
+Every operator has many kernels. It because there are multiple data types, places, data layout that Fluid supports. We use the `KernelType` to describe kernel types that operators can hold.
```
Comment (QiJune) [inline doc/design/switch_kernel.md]: It --> It's

File: doc/design/switch_kernel.md (hunk 3 above)
Comment (typhoonzero) [reply doc/design/switch_kernel.md]: I see, thank you guys for detailed explains.

File: doc/design/switch_kernel.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+We register a kernel for every operator and every kernel type ideally. However, it is impracticable for the following situations.
+
+1. Some operators, like CRF, are complicated and inefficient to be implemented on GPU. The CRF operator will only have a CPU kernel.
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed in GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
```
Comment (QiJune) [inline doc/design/switch_kernel.md]: > The inputs of this operator should be run on kernel type
Since only operator will have kernel type,  do you mean `kernel type` here is type of the previous operator that generates the inputs ?

File: doc/design/switch_kernel.md (hunk 6 above)
Comment (tensor-tang) [reply doc/design/switch_kernel.md]: The reason why I wrote like this is: (maybe it's just my concern)
1. the expected layout that are determined by current setting (eg. conv settings) and the platform (AVX2 or AVX512) **inside** the mkldnn kernel.

2. Is that possible that layout itself could be an attribute of input tensor not this kernel, so in the device who want to change the layout can just get from input tensor. No need to register may kernel types?

File: doc/design/switch_kernel.md (hunk 1 above)
Comment (tensor-tang) [reply doc/design/switch_kernel.md]: Thx @reyoung , indeed I think its another topic.

Looking forward to that.

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (typhoonzero) [reply doc/design/switch_kernel.md]: > There are many operators will be implemented by MKL. OpA --> OpB --> OpC. If OpB and OpC use the MKL optimized layout format, we just transform data in OpB.


"DataLayout" is one attribute of the typed data (or LoDTensor which also have lod as its attribute), in the [Dataflow](https://en.wikipedia.org/wiki/Dataflow_programming) model, we prefer less states of the kernel. For efficiency, kernels do not need to transform the layout if it is already satisfied.

File: doc/design/switch_kernel.md (hunk 4 above)
Comment (jacquesqiao) [inline doc/design/switch_kernel.md]: Do we need to support different kind of kernel with the same KernelType, for example, one CPU kernel is written in native cpp while another one is written with some Blas library.  If we will support, then how to get the right kernel

File: doc/design/switch_kernel.md (hunk 7 above)
Comment (jacquesqiao) [reply doc/design/switch_kernel.md]: > transforming inputs of an operator from a specific kernel type to another kernel type

`transforming inputs of an operator to support a certain kernel type`?

File: doc/design/switch_kernel.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+## Solution
+
+It is clearly that transforming inputs of an operator from a specific kernel type to another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
+
+We can infer a kernel type from the inputs of an operators. We let this kernel type as `actual kernel type`, which means this kernel type is the actually kernel type that operator should be performed.
```
Comment (jacquesqiao) [inline doc/design/switch_kernel.md]: we let ==> we define

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (jacquesqiao) [inline doc/design/switch_kernel.md]: should Executor has Place as it's property?

File: doc/design/switch_kernel.md (hunk 6 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: 1. The layout is an attribute of input tensor.
2. Switching kernel is not a job of the kernel itself. In Fluid, the kernel has only one method, `Kernel::Compute`. Switching kernel is not only used by `Conv` operator which uses `MKLDNN` as its backend but also used by `Pool(MKLDNN backend)` operator and many other operators. It is a higher level job. Because operators get and invoke `Kernel::Compute`. It is reasonable to let operators handle kernel switching.
3. `No need to register may kernel types?` All of these kernel types can be registered by a template class. For developers, they can just write one class and register it to every kernel type.
4. > the expected layout are determined by current setting (eg. conv settings) and the platform (AVX2 or AVX512) inside the mkldnn kernel.

    Yes. If Fluid is compiled with MKLDNN, the logic `GetExpectedKernelType` of `conv` and `pool` is just like you wrote.

File: doc/design/switch_kernel.md (hunk 8 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: It is another topic.

However, operator developers may be familiar with plain C++, not a register syntax.

File: doc/design/switch_kernel.md (hunk 10 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: Yes. 

1. The input data is generated by previous operators.
2. Operators can infer the actual kernel type from its inputs.

So operators can infer the actual kernel type from the outputs of previous operators.

File: doc/design/switch_kernel.md (hunk 4 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: No, we do not.

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (reyoung) [reply doc/design/switch_kernel.md]: Not sure.

However, this design will fit any situation whether executor has or does not have Place.

File: doc/design/switch_kernel.md (hunk 4 above)
Comment (jacquesqiao) [reply doc/design/switch_kernel.md]: if a user wants to add a custom CPU kernel that is already implemented and registered, how should he register and choose to use it? The data type, place, and layout are all the some.

File: doc/design/switch_kernel.md (hunk 2 above)
Comment (jacquesqiao) [reply doc/design/switch_kernel.md]: ok~

File: doc/design/switch_kernel.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+2. Some operators will take too many memory. It is better to force them into CPU. However, the rest of operators in this neural network will be performed on GPU, i.e., model parallel problem.
+3. Some layout and place are particular. One example is that MKLDNN uses `nChw8` and there is no other library uses `nChw8c`.
+
+Problems under these situations are similar. We can formalise this problem as follow.
+
+We register kernels with types $KT = \{kt_1, kt_2, kt_3, ...\}$ for one operator. The inputs of this operator should be run on kernel type $kt_{?}$, which the $kt_{?} \notin KT$. How to cast the input of this operator from $kt_{?}$ to any of kernel type in $KT$.
+
+## Solution
+
+It is clearly that transforming inputs of an operator toadapt another kernel type is not related to the particular operator. So we should register these transformation methods as global methods.
```
Comment (gongweibao) [inline doc/design/switch_kernel.md]: toadapt => to adapt
